include/pmacc/dimensions/DataSpaceOperations.hpp

```diff
@@ -165,10 +166,11 @@ namespace pmacc
 
         static HDINLINE DataSpace<DIM2> map(const DataSpace<DIM2>& size, uint32_t pos)
         {
-            return DataSpace< DIM2 >(
-                pos % size.x(),
-                pos / size.x()
-            );
+            auto const y = pos / size.x();
+            // pos is the resulting x
+            pos -= y * size.x();
```

Comment:
shouldn't an
```C++
auto const x = pos - y * size.x()
```

be free as well?

Reply:
I tried it and it is a little bit more efficient (use less register) than my initial version. Explicit compiler optimizations for the win :-)

Reply:
Lol, alright

Reply:
CNA you post the before/after registers? I am curious

Reply:
Here is one kernel with significant reduced registers (note that this not automatically mean that we get a performance increase):

```
# before ax3l suggestion 21 register are used
[0mptxas info    : Function properties for _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEjN5cupla11CuplaKernelIN5pmacc6nvidia6reduce6kernel6ReduceINS8_4math6VectorIfLi3ENSD_16StandardAccessorENSD_17StandardNavigatorENSD_6detail17Vector_componentsEEELj8ELj8EEEEEJNS8_17DataBoxDim1AccessINS8_7DataBoxINS8_10PitchedBoxISJ_Lj2EEEEEEEjPSJ_NS9_8functors3MaxENST_6AssignEEEEvNS_3vec3VecIT_T0_EET1_DpT2_[0m
[0m    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads[0m
[0mptxas info    : Used 21 registers, 386 bytes cmem[0][0m

# after 17 register are used
[0mptxas info    : Compiling entry function '_ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEjN5cupla11CuplaKernelIN5pmacc6nvidia6reduce6kernel6ReduceINS8_4math6VectorIfLi3ENSD_16StandardAccessorENSD_17StandardNavigatorENSD_6detail17Vector_componentsEEELj8ELj8EEEEEJNS8_17DataBoxDim1AccessINS8_7DataBoxINS8_10PitchedBoxISJ_Lj2EEEEEEEjPSJ_NS9_8functors3MaxENST_6AssignEEEEvNS_3vec3VecIT_T0_EET1_DpT2_' for 'sm_35'[0m
[0mptxas info    : Function properties for _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEjN5cupla11CuplaKernelIN5pmacc6nvidia6reduce6kernel6ReduceINS8_4math6VectorIfLi3ENSD_16StandardAccessorENSD_17StandardNavigatorENSD_6detail17Vector_componentsEEELj8ELj8EEEEEJNS8_17DataBoxDim1AccessINS8_7DataBoxINS8_10PitchedBoxISJ_Lj2EEEEEEEjPSJ_NS9_8functors3MaxENST_6AssignEEEEvNS_3vec3VecIT_T0_EET1_DpT2_[0m
[0m    0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads[0m
[0mptxas info    : Used 17 registers, 386 bytes cmem[0][0m
```

Reply:
Wow. NVCC, we need to talk.
CUDA8.0 or 9.0?

Reply:
human redable name of the kernel
```
void alpaka::exec::cuda::detail::cudaKernel<std::integral_constant<unsigned long,
    3ul
>,
    unsigned int,
    cupla::CuplaKernel<pmacc::nvidia::reduce::kernel::Reduce<pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>,
    8u,
    8u
> 
>,
    pmacc::DataBoxDim1Access<pmacc::DataBox<pmacc::PitchedBox<pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>,
    2u
> 
> 
>,
    unsigned int,
    pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>*,
    pmacc::nvidia::functors::Max,
    pmacc::nvidia::functors::Assign
>(alpaka::vec::Vec<std::integral_constant<unsigned long,
    3ul
>,
    unsigned int
>,
    cupla::CuplaKernel<pmacc::nvidia::reduce::kernel::Reduce<pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>,
    8u,
    8u
> 
>,
    pmacc::DataBoxDim1Access<pmacc::DataBox<pmacc::PitchedBox<pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>,
    2u
> 
> 
>,
    unsigned int,
    pmacc::math::Vector<float,
    3,
    pmacc::math::StandardAccessor,
    pmacc::math::StandardNavigator,
    pmacc::math::detail::Vector_components
>*,
    pmacc::nvidia::functors::Max,
    pmacc::nvidia::functors::Assign)
```

Reply:
It was checked with CUDA8   (CUDA9 is currently not available on our cluster :-()

Reply:
Maybe try proxygon! :)

Reply:
I can but there is no need. The initial goal of this PR is to remove the modulo.

Reply:
Sure, I was just wondering if nvcc 9 is still as ... interesting with the *second* change :)